core/src/main/java/io/grpc/Metadata.java

```diff
@@ -132,87 +135,145 @@ public Integer parseAsciiString(String serialized) {
+    return namesAndValues != null ? namesAndValues.length : 0;
+  }
+
+  // The scaled version of size.
+  private int len() {
+    return size * 2;
   }
 
   /**
    * Constructor called by the application layer when it wants to send metadata.
    */
   public Metadata() {}
 
-  private void storeAdd(String name, MetadataEntry value) {
-    List<MetadataEntry> values = store.get(name);
-    if (values == null) {
-      // We expect there to be usually unique header values, so prefer smaller arrays.
-      values = new ArrayList<MetadataEntry>(1);
-      store.put(name, values);
-    }
-    storeCount++;
-    values.add(value);
-  }
-
   /**
    * Returns the total number of key-value headers in this metadata, including duplicates.
    */
   @Internal
   public int headerCount() {
-    return storeCount;
+    return size;
   }
 
   /**
    * Returns true if a value is defined for the given key.
    */
   public boolean containsKey(Key<?> key) {
-    return store.containsKey(key.name());
+    for (int i = 0; i < size; i++) {
+      if (Arrays.equals(key.asciiName(), name(i))) {
+        return true;
+      }
+    }
+    return false;
   }
 
   /**
    * Returns the last metadata entry added with the name 'name' parsed as T.
    * @return the parsed metadata entry or null if there are none.
    */
   public <T> T get(Key<T> key) {
-    List<MetadataEntry> values = store.get(key.name());
-    if (values == null) {
-      return null;
+    for (int i = size - 1; i >= 0; i--) {
+      if (Arrays.equals(key.asciiName(), name(i))) {
+        if (typedValues[i] == null) {
+          typedValues[i] = key.parseBytes(value(i));
+        }
+        @SuppressWarnings("unchecked") // take THAT typesafety!
+        T ret = (T) typedValues[i];
+        return ret;
+      }
     }
-    MetadataEntry metadataEntry = values.get(values.size() - 1);
-    return metadataEntry.getParsed(key);
+    return null;
   }
 
   /**
    * Returns all the metadata entries named 'name', in the order they were received,
    * parsed as T or null if there are none. The iterator is not guaranteed to be "live." It may or
    * may not be accurate if Metadata is mutated.
    */
-  public <T> Iterable<T> getAll(Key<T> key) {
-    if (containsKey(key)) {
-      /* This is unmodifiable currently, but could be made to support remove() in the future.  If
-       * removal support is added, the {@link #storeCount} variable needs to be updated
-       * appropriately. */
-      return new ValueIterable<T>(key, store.get(key.name()));
+  public <T> Iterable<T> getAll(final Key<T> key) {
+    for (int i = 0; i < size; i++) {
+      if (Arrays.equals(key.asciiName(), name(i))) {
+        final int k = i;
+        return new Iterable<T>() {
+          @Override
+          public Iterator<T> iterator() {
+            return new Iterator<T>() {
+              int lastIdx = k;
+
+              @Override
+              public boolean hasNext() {
+                return lastIdx < size;
```

Comment:
Won't calling hasNext() multiple times then mess up the values returned by next then?